\documentclass[10pt,letterpaper]{article}
\usepackage[top=0.85in,left=2.5in,footskip=0.75in]{geometry}

% amsmath and amssymb packages, useful for mathematical formulas and symbols
\usepackage{amsmath,amssymb}

% Use adjustwidth environment to exceed column width (see example table in text)
\usepackage{changepage}

% Use Unicode characters when possible
\usepackage[utf8x]{inputenc}

% textcomp package and marvosym package for additional characters
\usepackage{textcomp,marvosym}

% color can be used to apply background shading to table cells only
\usepackage[table]{xcolor}

% Use nameref to cite supporting information files (see Supporting Information section for more info)
\definecolor{linkblue}{RGB}{0,76,153}
\usepackage[colorlinks=true,linkcolor=linkblue,citecolor=linkblue,urlcolor=linkblue]{hyperref}
\usepackage{nameref}

% line numbers
\usepackage[right]{lineno}

% ligatures disabled
\usepackage{microtype}
\DisableLigatures[f]{encoding = *, family = * }

% array package and thick rules for tables
\usepackage{array}

% create "+" rule type for thick vertical lines
\newcolumntype{+}{!{\vrule width 2pt}}

% create \thickcline for thick horizontal lines of variable length
\newlength\savedwidth
\newcommand\thickcline[1]{%
  \noalign{\global\savedwidth\arrayrulewidth\global\arrayrulewidth 2pt}%
  \cline{#1}%
  \noalign{\vskip\arrayrulewidth}%
  \noalign{\global\arrayrulewidth\savedwidth}%
}

% \thickhline command for thick horizontal lines that span the table
\newcommand\thickhline{\noalign{\global\savedwidth\arrayrulewidth\global\arrayrulewidth 2pt}%
\hline
\noalign{\global\arrayrulewidth\savedwidth}}

% additional packages imported by Jesse Bloom
\usepackage[export]{adjustbox}
\usepackage{float}

% added by Jesse Bloom for full-width figures
\newenvironment{fullwidth}{%
  \begin{adjustwidth}{-2.05in}{}
}{\end{adjustwidth}}

% Added by Jesse Bloom for figures
\renewcommand{\floatpagefraction}{0.99}
\renewcommand{\textfraction}{0.01}
\setcounter{topnumber}{8}
\setcounter{bottomnumber}{8}
\setcounter{totalnumber}{8}

\newfloat{suppfile}{thp}{lofsupfile}
\floatname{suppfile}{File}
\renewcommand{\thesuppfile}{S\arabic{suppfile}}

\newfloat{suppfig}{thp}{lofsupfile}
\floatname{suppfig}{Figure}
\renewcommand{\thesuppfig}{S\arabic{suppfig}}

\renewcommand{\figureautorefname}{Figure}
\newcommand{\FIG}[1]{\autoref{fig:#1}}

\newcommand{\suppfileautorefname}{File}
\newcommand{\SUPPFILE}[1]{\autoref{suppfile:#1}}

\newcommand{\suppfigautorefname}{Figure}
\newcommand{\SUPPFIG}[1]{\autoref{suppfig:#1}}

% Remove comment for double spacing
%\usepackage{setspace} 
%\doublespacing

% Text layout
%\raggedright
\setlength{\parindent}{0.5cm}
\textwidth 5.5in 
\textheight 8.75in

% Bold the 'Figure #' in the caption and separate it from the title/caption with a period
% Captions will be left justified
\usepackage[aboveskip=1pt,labelfont=bf,labelsep=period,justification=raggedright,singlelinecheck=off]{caption}
\renewcommand{\figurename}{Figure}

% Harvard-style citations
\usepackage{natbib}
\bibliographystyle{cell}

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother


% Header and Footer with logo
\usepackage{lastpage,fancyhdr,graphicx}
\usepackage{epstopdf}
%\pagestyle{myheadings}
\pagestyle{fancy}
\fancyhf{}
%\setlength{\headheight}{27.023pt}
%\lhead{\includegraphics[width=2.0in]{PLOS-submission.eps}}
\rfoot{\thepage/\pageref{LastPage}}
\renewcommand{\headrulewidth}{0pt}
\renewcommand{\footrule}{\hrule height 2pt \vspace{2mm}}
\fancyheadoffset[L]{2.25in}
\fancyfootoffset[L]{2.25in}
\lfoot{\today}

% section header formats
\usepackage{sectsty}
\definecolor{darkred}{RGB}{140,0,0}
\sectionfont{\normalsize \color{darkred}}
\subsectionfont{\normalsize \color{darkred}}
\subsubsectionfont{\normalsize \color{darkred}}
\paragraphfont{\normalsize \color{darkred}}

%% Include all macros below

\providecommand{\keywords}[1]{\noindent \textbf{Keywords:} #1}

%% END MACROS SECTION


\begin{document}
\vspace*{0.2in}

% Title must be 250 characters or less.
\begin{flushleft}
{\LARGE
\textbf{Single-cell virus sequencing of influenza infections that trigger innate immunity}}
\newline
% Insert author names, affiliations and corresponding author email (do not include titles, positions, or degrees).
\\
{\bf
Alistair B. Russell\textsuperscript{1},
Jacob R. Kowalsky\textsuperscript{2},
Jesse D. Bloom\textsuperscript{1,2,3*}
}
\\
\bigskip
{\small
$^1$Basic Sciences Division and Computational Biology Program, Fred Hutchinson Cancer Research Center
\\
$^2$Department of Genome Sciences, University of Washington
\\
$^3$Howard Hughes Medical Institute
\\
Seattle, WA  98109, USA
\\
*Correspondence: jbloom@fredhutch.org
}
\bigskip


\end{flushleft}

\section*{SUMMARY}
{\bf
The outcome of viral infection is extremely heterogeneous, with infected cells only sometimes activating innate immunity.  
Here we develop a new approach to assess how the genetic variation inherent in viral populations contributes to this heterogeneity.
We do this by determining both the transcriptome and full-length sequences of all viral genes in single influenza-infected cells.
Most cells are infected by virions with defects such as amino-acid mutations, internal deletions, or failure to express a gene.
We identify instances of each type of defect that increase the likelihood that a cell activates an innate-immune response.
However, immune activation remains stochastic in cells infected by virions with these defects, and sometimes occurs even when a cell is infected by a virion that expresses unmutated copies of all genes.
Our work shows that viral genetic variation substantially contributes to but does not fully explain the heterogeneity in single influenza-infected cells.
}

\vspace{0.2in}
\keywords{PacBio, single-cell RNA-seq, influenza, interferon, NS1, stochastic}

\linenumbers

% Use "Eq" instead of "Equation" for equation citations.
\section*{INTRODUCTION}
Infection with an acute virus such as influenza initiates a race between the virus and immune system.
As the virus spreads, some cells detect infection and produce interferon (IFN).
This IFN directs expression of anti-viral interferon-stimulated genes (ISGs) in the infected cell and its neighbors via autocrine and paracrine signaling, as well as helping launch a broader immune response~\citep{stetson2006type,honda2006type}.
If innate immunity is activated sufficiently rapidly, it can reduce viral replication and disease~\citep{solov1969results,treanor1987intranasally,beilharz2007protection,kugel2009intranasal,steel2010transmission}---although excessive immune responses later in infection can actually be associated with immunopathology and severe disease~\citep{la2007question, iwasaki2014innate}.

Unfortunately for the host, influenza initially only rarely triggers IFN production by infected cells~\citep{kallfass2013visualizing, killip2017single}.
This rareness of IFN induction is just one form of the extreme cell-to-cell heterogeneity that characterizes infection: cells also vary widely in their production of viral mRNA, proteins, and progeny virions~\citep{russell2018extreme,steuerman2018dissection,sjaastad2018distinct,heldt2015single}.
Because viral growth and the IFN response are both feed-forward processes, early cell-to-cell heterogeneity could have significant downstream consequences for the race between virus and immune system---especially since natural human infections are typically initiated by just a few virions entering a few cells~\citep{mccrone2018stochastic, xue2018reconciling, varble2014influenza}.

It is unclear why only some infected cells trigger innate-immune responses.
Two possible contributors are pure stochasticity and pre-existing variation in cellular state.
For instance, only some cells induce IFN even upon treatment with synthetic innate-immune ligands~\citep{shalek2013single, shalek2014single, wimmers2018single}, and the frequency of IFN induction may depend on a cell's pre-existing chromatin state~\citep{bhushal2017cell}.
But for influenza, a third possible contributor looms large: viral genetic diversity.
Because influenza has a high mutation rate, individual virions often have defects~\citep{parvin1986measurement, suarez1992heterogeneity, bloom2014experimentally, pauly2017novel}.
Indeed, many studies have identified mutations that increase IFN induction when engineered into a viral population~\citep{velthuis2018mini, du2018genome, killip2017single, perez2014unbiased}, and viral stocks that are rich in internal deletions in the polymerase genes induce more IFN~\citep{baum2010preference, tapia2013defective, boergeling2015evidence, dimmock2015cloned}.

However, existing techniques are inadequate to determine how viral genetic diversity contributes to cell-to-cell heterogeneity during infection.
Flow cytometry and fluorescent reporters only measure protein levels~\citep{brooke2013most, guo2017single}, and current single-cell transcriptomic techniques primarily measure abundance of transcripts and provide only fragmentary information on their sequences~\citep{russell2018extreme, zanini2018single, zanini2018virus, steuerman2018dissection, saikia2018simultaneous, oneal2018west}.
None of these techniques reliably reveal if there are mutations in the virion infecting any given single cell.

Here we develop a new approach to determine both the transcriptome and full sequences of all viral genes in single influenza-infected cells.
To do this, we perform both standard Illumina-based transcriptomics and full-length PacBio sequencing of viral genes from single cells.
We obtain transcriptomes and sequences of all expressed viral genes in 150 infected cells, 40 of which express IFN.
Two-thirds of cells are infected by virions with a mutation or defect in gene expression.
This viral diversity is a major contributor to cell-to-cell heterogeneity, with cells infected by unmutated virions having a tighter distribution of viral transcriptional burden.
We identify several types of viral defects that increase IFN induction.
However, viral genetic variation does not fully explain the heterogeneity, and even unmutated virions sometimes induce IFN.
Therefore, viral diversity is an important but not exclusive cause of cell-to-cell heterogeneity during influenza infection.

\section*{RESULTS}

\subsection*{A system to identify and enrich rare IFN+ cells}

%%% start IFNrare figure %%%
\begin{figure}
\centerline{
{\bf \Large A}\includegraphics[width=0.34\textwidth,valign=t]{figures/IFN_stochastic/IFN_reporter/IFN_reporter.pdf}
\hspace{0.02\textwidth}
{\bf \Large B} \includegraphics[width=0.22\textwidth,valign=t]{figures/IFN_stochastic/Flow/ifn_percent.pdf}
}
\caption{
{\bf Reporter cells to identify and enrich infections that activate IFN expression.}
{\bf (A)} The reporter consists of an IFN promoter that drives expression of a cell-surface protein amenable to MACS and a fluorescent protein.
We created reporters with type I and type III IFN promoters (\SUPPFILE{plasmids}).
In A549 cells, the reporters were efficiently activated by an IFN-inducing strain of Sendai virus (\SUPPFIG{reportervalidation}A).
{\bf (B)}
Frequency of IFN induction upon infection with the influenza virus stock used in the single-cell studies in this paper, as quantified using the type III IFN reporter (see \SUPPFIG{reportervalidation}C for full data).
The plot also shows uninfected cells, and cells infected with saturating amounts of Sendai virus.
The limit of detection of 0.05\% is indicated with a dashed line, and numbers show the median of three measurements.
}
\label{fig:IFNrare}
\end{figure}
%%% end IFNrare figure %%%

A challenge in studying IFN induction by influenza virus is its rareness at the level of single cells.
Studies using reporters~\citep{killip2017single,kallfass2013visualizing} and our own previous single-cell transcriptomics~\citep{russell2018extreme} have found that at early timepoints IFN induction occurs in $<$1\% of influenza-infected cells.
To efficiently identify and enrich these rare IFN+ cells, we created A549 cells that carried IFN reporters consisting of a type I (\textit{IFNB1}) or type III (\textit{IFNL1}) promoter driving expression of a cell-surface protein ~\citep[LNGFR$\Delta$C;][]{bonini1997hsv,ruggieri1997cell} followed by a fluorescent protein (\FIG{IFNrare}A).
Cells that activate the IFN reporter can be enriched by magnetic-activated cell sorting (MACS) or identified by flow cytometry.
The reporters were efficiently activated by infection with saturating amounts of a strain of Sendai virus~\citep{strahle2006sendai} that potently induces IFN (\SUPPFIG{reportervalidation}A), and activation of the type I and type III IFN reporters was highly correlated in our cells (\SUPPFIG{reportervalidation}B; further validated by the single-cell transcriptomics below).
For the rest of this paper, we use ``IFN expression'' to refer to combined expression of type I and III IFNs. 

We generated a stock of A/WSN/1933 (H1N1) influenza (hereafter referred to as ``WSN''), and found that it activated the IFN reporter in $\sim$0.5\% of infected cells  (\FIG{IFNrare}B), a frequency roughly comparable to that in our prior single-cell transcriptomics of influenza-infected cells~\citep{russell2018extreme}.


\subsection*{Combined transcriptomics and virus-sequencing of single infected cells}

%%% start workflow figure
\begin{figure}
\begin{fullwidth}

\includegraphics[width=\linewidth, valign=t]{figures/WorkflowSchematic/SchematicForPaper.pdf}

\caption{
{\bf Approach for combined transcriptomics and virus sequencing of single influenza-infected cells that express IFN.}
{\bf (A)}
IFN reporter A549 cells are infected with a mix of wild-type and synonymously barcoded viruses.
IFN+ cells are enriched by MACS (\SUPPFIG{MACS}), and pooled with non-enriched cells and uninfected canine cells that serve as an internal control for multiplets and mRNA leakage.
{\bf (B)}
The mRNAs from individual cells are converted to cDNAs tagged with cell-specific barcodes.
{\bf (C)}
Cellular transcriptomes are quantified using standard single-cell 3'-end Illumina sequencing, and 
{\bf (D)}
viral genes are enriched by influenza-specific PCR and fully sequenced by PacBio (in this schematic, only the cell labeled by the red barcode is infected and has viral transcripts that are enriched and sequenced by PacBio).
{\bf (E)}
The result is a matrix giving the expression of each gene in each cell, as well as the full sequences of the viral genes in infected cells.
}
\label{fig:workflow}

\end{fullwidth}
\end{figure}
%%% end workflow figure

To determine if mutations or other defects in the infecting virions contribute to the heterogeneous outcome of infection, we developed the approach in \FIG{workflow} to obtain the entire transcriptome \emph{and} the full sequences of all viral genes in single cells.
First, we generated a stock of virus that consisted of a mix of wild-type WSN and a ``synonymously barcoded'' variant that contained two engineered synonymous mutations near each termini of each gene (\SUPPFILE{plasmids}).
These viral barcodes allow us to identify co-infections~\citep{russell2018extreme}, and provide a control for PCR artifacts during full-length sequencing of viral transcripts (see below).
We used this viral stock to infect A549 IFN reporter cells (\FIG{workflow}A) at a dose that led to detectable viral transcription in about a quarter of cells.
From 12 to 13 hours post-infection, we used MACS to enrich cells that activated the IFN reporter (\SUPPFIG{MACS}).
To ensure the presence of IFN- cells, we added back non-enriched cells to $\sim$10\% of the total.
We also added uninfected canine cells to $\sim$5\% of the total as a control for multiplets and to estimate the background amount of viral mRNA detected in truly uninfected cells.

We processed the cells on a commercially available platform~\citep{zheng2017massively} that isolates cells in droplets and reverse transcribes polyadenylated mRNAs to append a unique cell barcode to all cDNAs in each droplet, and a unique molecular identifier (UMI) to each cDNA molecule (\FIG{workflow}B).
Because influenza virus mRNAs are polyadenylated~\citep{robertson1981polyadenylation}, this process appends cell barcodes to both cellular and viral mRNAs.
Furthermore, because virtually the entire influenza genome is transcribed, the cell-barcoded cDNA spans almost all 13,581 nucleotides in the segmented viral genome: the only portions not covered are one universally conserved nucleotide upstream of the transcription start site~\citep{koppstein2015sequencing} and 17 to 22 highly conserved nucleotides downstream of the polyadenylation site~\citep{robertson1981polyadenylation} in each of the eight viral gene segments.

We used a portion of the cell-barcoded cDNA for standard single-cell transcriptomics by Illumina 3'-end sequencing (\FIG{workflow}C).
But we also took a portion and enriched for full-length viral molecules by PCR (\FIG{workflow}D).
We performed PacBio sequencing on these full-length viral cDNAs to generate high-accuracy circular consensus sequences~\citep[CCSs;][]{travers2010flexible}.
These CCSs retain the cell barcodes, and with sufficient sequencing depth we obtain CCSs from multiple unique UMI-tagged cDNAs for each viral gene in each cell.
Because most cells are infected by just one or two virions, we can build a consensus of CCSs for each viral gene in each cell to determine the sequence(s) of these virions.
Combining this information with the 3'-end sequencing determines the entire transcriptome and the full sequences of the infecting virions in single cells (\FIG{workflow}E).

\subsection*{Transcriptomic analyses of single IFN+ and IFN- influenza-infected cells}

%%% start transcriptomics figure
\begin{figure}
\begin{fullwidth}

\includegraphics[width=\linewidth, clip=false]{figures/single_cell_figures/p_cell_summary.pdf}

\caption{
{\bf Single-cell transcriptomics of IFN-enriched influenza-infected cells.}
{\bf (A)} 
Number of cells for which transcriptomes were obtained.
%From these numbers, we estimate~\citep{bloom2018estimating} that $\approx$11\% of the transcriptomes are actually multiplets derived from multiple cells. 
{\bf (B)} The number of cellular and viral mRNAs detected for each cell is plotted as a point.
Green lines show the distribution of cellular mRNAs per cell.
Cells outside the dashed magenta lines have unusually low or high amounts of cellular mRNA (likely low-quality emulsions or multiplets), and are excluded from subsequent analyses.
{\bf (C)} Distribution across cells of the fraction of all mRNA derived from influenza.
Cells called as infected are in blue, while other cells are in green.
The inset shows the amount of viral mRNA in the human cells that are called as infected.
{\bf (D)} Number of influenza genes detected per infected cell, and the amount of viral mRNA in cells expressing each number of viral genes.
The majority of cells express all eight gene segments, but a substantial minority fail to express at least one gene.
\SUPPFIG{transcriptomics_supp}A shows the frequency that each viral gene is detected.
{\bf (E)} Relative expression of viral genes, quantified as the fraction of all viral mRNA in each infected cell that is derived from each gene.
{\bf (F)} Number of cells infected with wild-type virus, synonymously barcoded virus, or both.
From the cells infected with both viral barcodes, we estimate~\citep{bloom2018estimating} that 63\% of infected cells are co-infected.
{\bf (G)} Fraction of cellular mRNA from IFN across cells, faceted by whether the cells are infected.
Cells to the left of the first dashed magenta line are classified as IFN-, and cells to the right of the second line are classified as IFN+.
A pseudocount is added to the number of IFN transcripts detected in each cell, which is why none of the fractions are zero.
Many cells that do not express IFN still express ISGs (\SUPPFIG{transcriptomics_supp}C,D).
}
\label{fig:transcriptomics}

\end{fullwidth}
\end{figure}
%%% end transcriptomics figure

We obtained transcriptomes for 1,614 human (A549) cells, and 50 of the uninfected canine cells that were spiked into the experiment as a control (\FIG{transcriptomics}A).
We also obtained 12 transcriptomes with a mix of human and canine transcripts; from the number of such mixed cell-type transcriptomes, we estimate~\citep{bloom2018estimating} that $\sim$11\% of the transcriptomes are derived from multiple cells.
To remove some of these multiplets along with low-quality droplets, we filtered transcriptomes with unusually high or low numbers of cellular transcripts as is commonly done in analysis of single-cell RNA-seq data~\citep{haque2017practical}.
After this filtering, we retained 1,490 human cells for further analysis (\FIG{transcriptomics}B).

To identify infected cells, we examined the fraction of each transcriptome derived from virus (\FIG{transcriptomics}C).
As expected, only a small fraction ($\sim$0.7\%) of transcripts in the uninfected canine cells were from influenza; this low-level background is from lysed cells that release ambient viral mRNA that is captured in droplets prior to reverse-transcription.
We tested whether each cell contained more viral transcripts than expected under a Poisson model given this background fraction, and classified 290 human cells as definitively infected by virus (\FIG{transcriptomics}C).
We classified the other cells as uninfected, although it is possible that some were infected with virions that produced very little mRNA.
The distribution of the amount of viral mRNA across infected cells is in the inset in \FIG{transcriptomics}C.
As in our prior work~\citep{russell2018extreme}, the distribution is extremely heterogeneous: many infected cells have only a few percent of mRNA derived from virus, but viral mRNA comprises over half the transcriptome of a few cells.

We called the presence or absence of each viral gene in each infected cell, again using a Poisson model parameterized by background fractions estimated from uninfected canine cells.
%We called presence / absence of genes rather than transcripts, since the two influenza genes that encode multiple transcripts (M1 / M2 from the M gene, and NS1 / NS2 from the NS gene) do so via alternative splicing that leaves both isoforms with the same termini, making them indistinguishable by 3'-end sequencing.
\FIG{transcriptomics}D (top panel) shows that a slight majority (162 of 290) of infected cells express all eight genes (see \SUPPFIG{transcriptomics_supp}A for frequencies for individual genes).
This measured frequency of infected cells expressing all eight genes is slightly higher than in our prior work~\citep{russell2018extreme} and studies by others~\citep{brooke2013most, heldt2015single, dou2017analysis}, which estimated that only 13\% to 50\% of infected cells express all genes. 

The amount of viral mRNA tended to be lower in cells that failed to express viral genes (\FIG{transcriptomics}D, bottom).
However, viral burden remained variable even after conditioning on the number of viral genes: some cells that failed to express one or even two genes still derived $>$50\% of their mRNA from virus, while other cells that expressed all genes had only a few percent of their mRNA from virus.
Consistent with prior work~\citep{russell2018extreme}, despite the wide variation in absolute expression of viral genes, their \emph{relative} expression was fairly consistent (\FIG{transcriptomics}E) and similar to values from older bulk studies~\citep{hatada1989control}.

By examining the synonymous viral barcodes near the 3' termini of transcripts, we determined that 38\% of cells were co-infected with wild-type and synonymously barcoded virions (\FIG{transcriptomics}F; cells called as co-infected if a binomial test rejected null hypothesis that $\ge$95\% of viral mRNA is from one viral barcode variant).
From \FIG{transcriptomics}F, we estimate~\citep{bloom2018estimating} that 63\% of infected cells are co-infected, implying that 25\% are co-infected with two virions with the same viral barcode (such co-infections cannot be identified from transcriptomic data).
This co-infection rate is higher than expected from the relative numbers of infected and uninfected cells (\FIG{transcriptomics}C) if infection is Poisson.
This discrepancy could arise if the MACS for IFN+ cells also enriches co-infected cells, if infection is not truly Poisson, or if co-infection increases the likelihood that we identify a cell as infected given the thresholds in \FIG{transcriptomics}C.
This moderately high rate of co-infection may also explain why more cells in our experiment express all eight viral genes compared to some prior studies, as a co-infecting virion can complement a missing viral gene. 

We next examined expression of IFN and ISGs (\FIG{transcriptomics}G and \SUPPFIG{transcriptomics_supp}B-D).
Over 20\% of infected cells were IFN+ given the heuristic thresholds in \FIG{transcriptomics}G, indicating that the MACS enriched IFN+ cells far beyond their initial frequency of $\sim$0.5\% (\FIG{IFNrare}B) among infected cells.
Few ($\sim$1.3\%) uninfected cells were IFN+; the few that were present might be because the MACS enriched for rare cells that spontaneously activated IFN, or because some cells that we classified as uninfected were actually infected at low levels.
Many more cells expressed ISGs than IFN itself: the ratio of ISG+ to IFN+ cells was 1.8 among infected cells, and 7 among uninfected cells (\SUPPFIG{transcriptomics_supp}C).
The IFN+ cells were a subset of the ISG+ cells: IFN+ cells always expressed ISGs, but many ISG+ cells did not express IFN (\SUPPFIG{transcriptomics_supp}D).
These results are consistent with the established knowledge that IFN is expressed only in cells that directly detect infection, but that ISGs are also expressed via paracrine signaling in other cells~\citep{stetson2006type,honda2006type}.

\subsection*{Full genotypes of viruses infecting single IFN+ and IFN- cells}

%%% start genotypes figure %%%
\begin{figure}
\begin{fullwidth}
{\centering
\includegraphics[height=0.92\textheight]{figures/single_cell_figures/p_genotypes.pdf}
}
\caption{
{\bf Viral genotypes and infection outcomes in single cells.}
Arrows indicate presence of a viral gene from one (light blue) or both viral barcode variants (dark blue).
Circles and boxes indicate mutations or indels as described in the legend, with circle areas and box heights proportional to the fraction of CCSs with that mutation.
For dual-barcode infections, mutations / indels for the wild-type and synonymously barcoded viral variants are shown in the top and bottom half of the arrows, respectively. 
Green boxes at left show the percent of all mRNA in that cell derived from virus.
Orange boxes show the percent of cellular mRNA derived from IFN, with orange box frames indicating cells classified as IFN+ in \FIG{transcriptomics}G.
}
\label{fig:genotypes}

\end{fullwidth}
\end{figure}
%%% end genotypes figure %%%

We next used PacBio sequencing (\FIG{workflow}D) to determine the full sequences of the viral genes expressed in single infected cells.
Using PCR enrichment (see Methods), we obtained over 200,000 high-quality PacBio CCSs that mapped to an influenza gene and contained a cell barcode and UMI (\SUPPFIG{PacBio_supp}A).
Crucially, the synonymous viral barcodes at both termini of each gene enabled us to confirm that PCR strand exchange was rare (\SUPPFIG{PacBio_supp}B), meaning that the vast majority of CCSs correctly link the sequence of the viral transcript to cell barcodes and UMIs that identify the cell and molecule of origin.

We called a mutation in a viral gene if it was found in least two CCSs originating from different mRNAs (unique UMIs) and at least 30\% of all CCSs for that gene in that cell.
For cells co-infected with both viral barcode variants, we called mutations separately for each viral variant.
This strategy reliably identifies mutations in virions that initiate infection of cells infected with at most one virion of each viral barcode variant ($\sim$75\% of infected cells), as well as high-abundance mutations in cells co-infected with multiple virions of the same viral barcode.
It will not identify mutations that arise within a cell after the first few rounds of viral genome replication, since such mutations will not reach 30\% frequency in that cell.
Therefore, analogous to somatic variant calling in tumor sequencing~\citep{xu2014comparison, cibulskis2013sensitive}, there is a limit to our detection threshold: we cannot identify mutations that occur on just a small fraction of transcripts in a cell. 

We could call the sequences of all expressed viral genes in the majority of infected cells (\SUPPFIG{ncells}).
We were most effective at calling full viral genotypes in cells that expressed high amounts of viral mRNA and were infected by only one viral barcode variant (\SUPPFIG{ncells}).
But we also called full genotypes for many cells that had low viral burden or were co-infected by both viral barcode variants.

The 150 cells for which we called the full viral genotypes are shown in \FIG{genotypes}.
Simple visual inspection of this figure reveals a wealth of information.
For instance, the cell with the highest viral burden (\textit{cell 1} in \FIG{genotypes}, which has 65\% of its mRNA from virus) was infected by a virion that expressed wild-type copies all eight genes and did not induce detectable IFN.
But 12 of the other 13 cells with at least 50\% of their mRNA from virus were infected by virions that had a mutation or failed to express at least one gene, and five of these cells expressed IFN.
The two cells that expressed the most IFN (\textit{cell 13} and \textit{cell 123}) both lacked the viral NS gene, and many other IFN+ cells had different defects such as large internal deletions (e.g., \textit{cell 5} and \textit{cell 89}) or amino-acid mutations (e.g., \textit{cell 9}, \textit{cell 28}, and many others).
However, expressing unmutated copies of all eight genes did not guarantee a favorable outcome for the virus: for instance, the wild-type virion that infected \textit{cell 139} only managed to express viral mRNA to 6\% of the total transcriptome, and the wild-type virion that infected \textit{cell 105} still induced IFN.

\subsection*{Viral defects associated with infection outcome in single cells}

%%% begin mutations figure %%%
\begin{figure}
\begin{fullwidth}
{\centering
\includegraphics[width=\linewidth]{figures/single_cell_figures/p_mutations.pdf}
}
\caption{
{\bf Viral features associated with heterogeneity in infection outcome among cells for which we sequenced all expressed viral genes.}
{\bf (A)} 
Percent of all mRNA derived from virus, faceted by whether cells express unmutated copies of all eight genes.
Cells infected by fully wild-type virus exhibit less heterogeneity in viral burden as quantified by the Gini index (95\% confidence intervals are indicated).
{\bf (B)}
IFN expression among cells expressing unmutated copies of all genes, and among cells with mutations or missing genes.
{\bf (C)}
Specific viral defects associated with IFN induction.
The top panel show the percent of IFN- and IFN+ cells that fail to express each viral gene.
The middle and bottom panels show the percent of IFN- and IFN+ cells that have a deletion or amino-acid substitution in each gene, conditioned on the cell expressing that gene.
Numbers give P-values (Fisher's exact test) for rejecting the null hypothesis that percents are equal among IFN- and IFN+ cells. 
{\bf (D)}
There is no association between IFN induction and the amount of viral mRNA in cells that express NS, but viral burden is associated with IFN induction among cells that lack NS.
Throughout this figure, we only consider substitutions that are non-synonymous.
Insertions are not shown as a mutation type as they are extremely rare (\FIG{genotypes}).
}
\label{fig:mutations}

\end{fullwidth}
\end{figure}
%%% end mutations figure %%%

To systematically assess viral features associated with infection outcome, we divided the 150 cells in \FIG{genotypes} into those that expressed unmutated copies of all eight genes (disregarding synonymous mutations) and those that did not.
\FIG{mutations}A shows that the 49 cells infected by full wild-type viruses had a significantly tighter distribution of the amount of viral mRNA per cell than the other 101 cells as quantified by the Gini index~\citep{gini1921measurement}.
Therefore, viral defects are a major contributor to the heterogeneity in viral transcriptional burden among cells.

Cells infected by incomplete or mutated virions also expressed IFN more frequently than cells infected by full wild-type virions (\FIG{mutations}B), although this difference was not statistically significant ($P = 0.12$, Fisher's exact test).
However, some specific viral defects were significantly associated with IFN induction: absence of NS and amino-acid mutations in PB1 were significantly enriched in IFN+ cells, and amino-acid mutations in NS and deletions in HA were weakly enriched (\FIG{mutations}C).
Due to the low number of cells and large number of hypotheses in \FIG{mutations}C, the only trend that remained significant at a false discovery rate of 10\% was absence of NS.

One other interesting trend emerges from the single-cell data.
There is no difference in the amount of viral mRNA between IFN+ and IFN- cells that express NS---but among cells that lack NS, the IFN+ ones have significantly more viral mRNA (\FIG{mutations}D).

\subsection*{Validation that viral defects in single IFN+ cells often cause increased IFN induction}

%%% begin validation figure %%%
\begin{figure}

\centerline{\includegraphics[width=0.85\textwidth]{figures/Validation_Figure/ifn_plot.pdf}}
\caption{
{\bf Validation that IFN induction is increased by some of the mutations identified in the single-cell virus sequencing of IFN+ cells.}
{\bf (A)}
Percent of infected cells that become IFN+ after infection with a bulk stock of the indicated viral mutant, as determined using a reporter cell line.
The numbers indicate the median of four measurements for each viral mutant.
The limit of detection of 0.05\% is indicated with a dashed green line, and the median value for wild type is indicated with a dashed blue line.
Points are colored orange if the mutant virus stock induces IFN more frequently than the wild-type viral stock (one-sided t-test, $P < 0.01$), and blue otherwise.
{\bf (B)}
Similar to the first panel, but validates the increased IFN induction for a large internal deletion in the PB1 gene.
See \SUPPFIG{validationflow} and \SUPPFIG{delqPCR} for more detailed experimental data.
The experiments in the two panels were performed on different days, and so numerical values can be reliably compared within panels but not between panels.
}
\label{fig:validation}

\end{figure}
%%% end validation figure %%%

To test if the viral defects identified in single IFN+ cells play a causal role in triggering IFN expression, we used reverse genetics to generate bulk stocks of viruses with some of these defects.

The viral defect most strongly associated with IFN induction was failure to express the NS gene (\FIG{genotypes}, \FIG{mutations}C).
Although it is sometimes possible to use complementing cells to generate influenza viruses lacking a specific gene~\citep{fujii2003selective,marsh2007specific}, we were unable to generate viruses that lacked NS.
The NS gene encodes two proteins (NS1 and NS2), the first of which is influenza's primary innate-immune antagonist~\citep{garcia1998influenza, hale2008multifunctional}.
We therefore mimicked the effect of absence of NS by creating a mutant virus (which we will term ``NS1stop'') that had multiple stop codons early in the NS1 coding sequence.

The single-cell data also showed that amino-acid substitutions in the PB1 and NS genes were enriched in IFN+ cells (\FIG{genotypes}, \FIG{mutations}C), so we created mutant viruses with many of the amino-acid substitutions found in these genes among IFN+ cells.
Overall, we created six such point-mutant viruses: PB1-D27N, PB1-G206S, PB1-K279R, PB1-T677A, NS1-A122V, and NS2-E47G.

Finally, prior work has suggested that virions carrying internal deletions in the polymerase genes can induce higher levels of IFN~\citep{baum2010preference, tapia2013defective, boergeling2015evidence, dimmock2015cloned}.
Although such deletions are not significantly enriched among IFN+ cells in our single-cell data (\FIG{mutations}C), there is an IFN+ cell that has a deletion in PB1 spanning nucleotides 385 to 2163 (\textit{cell 5} in \FIG{genotypes}).
We therefore created a virus carrying this deletion, and propagated it in cells constitutively expressing PB1 protein.

We tested the rate of IFN induction by each viral stock using the reporter cells.
\FIG{validation} shows that five of the eight mutant viral stocks induced IFN more frequently than wild type.
The strongest IFN induction was by the NS1stop virus, but the PB1 internal deletion and three of the point-mutant viruses (PB1-D27N, PB1-T677A, and NS1-A122V) also induced IFN significantly more frequently than wild type.
The other three point mutants (PB1-G206S, PB1-K279R, and NS2-E47G) did not increase IFN induction---an unsurprising finding, since we also expect some mutations without an IFN-enhancing effect to be found in IFN+ cells by chance.
But overall, the results in \FIG{validation} validate that the viral defects in single IFN+ cells often cause increased IFN production.

\section*{DISCUSSION}
We have determined the full sequences of all viral genes expressed in single influenza-infected cells, and examined how viral mutations and gene-expression defects associate with infection outcome in each cell.
Methodologically, our major advance is to measure the \emph{genotypes} of viruses in addition to the abundance of viral components (i.e., transcripts, proteins, or progeny virions) as has been done by prior single-cell studies~\citep{russell2018extreme, zanini2018single, zanini2018virus, steuerman2018dissection, saikia2018simultaneous, oneal2018west, zhu2009growth, schulte2014single, akpinar2016high, heldt2015single, brooke2013most}.
Our method builds on the observation by that fragmentary viral genetic information can be obtained by more standard single-cell transcriptomic techniques~\citep{saikia2018simultaneous, zanini2018virus}.
To make this information complete, we have coupled single-cell transcriptomics with long-read PacBio sequencing of viral genes, a strategy analogous to that used by \citep{gupta2018single} to obtain full-length isoforms of some cellular genes in single cells.

This viral genetic information turns out to be crucial for understanding infection outcome and innate-immune induction.
Despite the fact that we used a low-passage viral stock generated from plasmids encoding a ``wild-type'' influenza genome, most infected cells do not express wild-type copies of all viral genes.
Although our study is certainly not the first to note that influenza has a high mutation rate~\citep{parvin1986measurement, suarez1992heterogeneity, suarez1994estimation, bloom2014experimentally, pauly2017novel} and sometimes fails to express genes~\citep{brooke2013most, heldt2015single, dou2017analysis, russell2018extreme}, it is the first to directly observe the full spectrum of these defects across single cells.
Visual inspection of \FIG{genotypes} shows how any experiment that does not sequence viral genes in single cells is averaging across a vast number of hidden viral defects.

It turns out that these viral defects substantially contribute to the heterogeneity observed in prior single-cell studies of influenza~\citep{russell2018extreme, steuerman2018dissection, heldt2015single, sjaastad2018distinct}, including helping determine if a cell triggers an IFN response.
We identified four types of viral defects in single IFN+ cells that we validated to increase IFN induction. 
Two types of defects---absence of the NS gene and amino-acid mutation to the NS1 protein---presumably impair NS1's well-known ability to antagonize innate-immune pathways~\citep{garcia1998influenza, hale2008multifunctional}.
The third type of defect, amino-acid mutations in the PB1 protein, is consistent with recent work by showing that mutations to the viral polymerase affect generation of mini-viral RNAs that activate the innate-immune sensor RIG-I~\citep{velthuis2018mini}.
Finally, we found an internal deletion in the PB1 gene that enhances IFN induction, consistent with a large body of prior work showing that such deletions are immunostimulatory~\citep{baum2010preference, tapia2013defective, boergeling2015evidence, dimmock2015cloned}.
Given the extensive prior work describing such deletions, it is perhaps surprising we did not identify more of them in our single IFN+ cells.
There may be several reasons: we used relatively pure viral stocks~\citep{xue2016propagation} at modest MOI; our experiments preferentially captured cells with higher viral transcriptional load; and most prior studies have used techniques that can detect large deletions but not subtle point mutations.
The relative importance of different types of viral defects for IFN induction likely varies across infection conditions, viral strains, and cell types---so it remains an open question as to which defects are most relevant for early immune detection of actual viral infections in humans.

But although viral variation is clearly important, it still only partially explains heterogeneity among influenza-infected cells.
We find substantial breadth in viral transcriptional burden and occasional IFN induction even among cells infected with fully wild-type virions.
Additionally, none of the viral defects in IFN+ cells induce IFN deterministically in our validation experiments---and the apparently most immunostimulatory defect (absence of NS) also occurs in multiple IFN- cells in our single-cell dataset.
Of course, our single-cell virus sequencing only calls mutations present in a cell at relatively high frequency, and so could overlook low-frequency mutations that contribute to IFN induction.
However, we suspect that stochasticity or pre-existing cellular states also affect the outcome of infection, since heterogeneity in IFN induction is present even among cells treated with synthetic innate-immune ligands~\citep{shalek2013single, shalek2014single, wimmers2018single, bhushal2017cell}, as well as for other viruses~\citep{oneal2018west}.

Perhaps the most intriguing question is how the heterogeneity that we have described ultimately affects the macroscopic outcome of infection.
Natural human influenza infections are established by just a handful of virions~\citep{mccrone2018stochastic, xue2018reconciling, varble2014influenza} that then undergo exponential growth, and early IFN responses are amplified by paracrine signaling~\citep{stetson2006type,honda2006type}.
It is therefore plausible that genetic variation in the founding viral population could have a large effect on downstream events.
Extending our approaches to more complex systems could shed further light on how viral genetic variation interacts with cell-to-cell heterogeneity to shape the race between virus and immune system.

\small
\subsection*{ACKNOWLEDGMENTS}
We thank Cole Trapnell, Jason Underwood, Robert Bradley, Daniel Stetson, AJ Velthuis, Katherine Xue, and Hannah Itell for helpful suggestions.
We thank Andy Marty and the Fred Hutch Genomics Core for performing the deep sequencing.
This work was supported by the NIAID of the NIH under grant R01 AI127893 and a Burroughs Wellcome Fund Young Investigator in the Pathogenesis of Infectious Diseases grant to JDB.
ABR was supported by a postdoctoral fellowship from the Damon Runyon Cancer Research Foundation.
JRK was supported by a Washington Research Foundation Undergraduate Research Fellowship and a Mary Gates undergraduate research scholarship from the University of Washington.
JDB is an Investigator of the Howard Hughes Medical Institute.

\subsection*{AUTHOR CONTRIBUTIONS}
ABR and JDB designed the study.
ABR performed the experiments with assistance from JRK.
ABR and JDB computationally analyzed the data and wrote the paper.

\subsection*{DECLARATION OF INTERESTS}
The authors declare no competing interests.

\bibliography{references}

\section*{METHODS}

\subsection*{IFN reporter cell lines}
We created IFN reporter variants of the A549 human lung epithelial cell line (\FIG{IFNrare}A).
The parental A549 cell line used to create these reporters was obtained from ATCC (CCL-185), and was tested as negative for mycoplasma contamination by the Fred Hutch Genomics Core and authenticated using the ATCC STR profiling service.
The cells were maintained in D10 media (DMEM supplemented with 10\% heat-inactivated fetal bovine serum, 2 mM L-glutamine, 100 U of penicillin / ml, and 100 $\mu$g of streptomycin / ml) at 37$^{\circ}$C and 5\% carbon dioxide.

To create the type I interferon reporters, a 1kb promoter region upstream of the human IFNB1 gene were cloned into the pHAGE2 lentiviral vector~\citep{oconnell2010lentiviral}, with a NotI site immediately downstream of the promoter serving as an artificial Kozak sequence. 
Downstream of this NotI site, each of the following reporter constructs was cloned: mCherry, mNeonGreen, and low-affinity nerve growth factor lacking the C-terminal signaling domain (LNGFR$\Delta$C)~\citep{bonini1997hsv,ruggieri1997cell} linked to mNeonGreen by a P2A linker~\citep{kim2011high}.
The sequence of the last of these constructs is provided in \SUPPFILE{plasmids}.

To create the type III interferon reporters, a 1.2kb region upstream of the human IL29 (IFNL1) gene was cloned into the pHAGE2 vector, with the native Kozak sequence retained at the 3' end. 
Downstream of this promoter we cloned LNGFR$\Delta$C linked to ZsGreen via a P2A linker.
The sequence of this construct is provided in \SUPPFILE{plasmids}.
 
We used these constructs to generate lentiviral vectors and transduce of A549 cells in the presence of 5 $\mu$g polybrene.
We then sorted single transduced cells and expanded them.
A portion of the expanded cells were tested for reporter activity by transfecting poly(I:C) (a potent agonist of the RIG-I pathway), and we retained clones with strong activation.
Importantly, the cells that we retained for further use were not the same portion that were tested by poly(I:C) treatment, but rather a separate split of the same population---this avoids any selection on the cells from transient activation of IFN.
For the dual type I / type III reporter used in \SUPPFIG{reportervalidation}B, a single-cell clone of the type III reporter cell line was transduced with the type I reporter bearing the mCherry fluorescent marker, and then isolated and propagated as a single cell clone for the other cell lines.
All reporter lines tested negative for mycoplasma contamination by the Fred Hutch Genomics Core.

\SUPPFIG{reportervalidation}A shows validation of the reporter cell lines using infection with saturating amounts of the Cantell strain of Sendai virus (obtained from Charles River Laboratories).
For detection of the cell-surface bound LNGFR$\Delta$C, cells were stained with PE-conjugated anti-LNGFR (CD271) antibody from Miltenyi Biotec.

\subsection*{Viruses for single-cell experiments}
We performed the single-cell experiments using the A/WSN/1933 (H1N1) strain of influenza virus.
We used both the wild-type virus and a variant of the virus where synonymous mutations were added within a few 100 nucleotides of each termini of each gene segment.
We have used a similar synonymous viral barcoding strategy in our prior single-cell work~\citep{russell2018extreme} as it allows us to detect about half of co-infected cells based on the expression of both viral barcode variants.
In the current work, we extended this approach by placing synonymous barcodes near \emph{both} termini of the gene segments in order to quantify strand exchange during PacBio sequencing (\SUPPFIG{PacBio_supp}B).
The sequences of all gene segments from the wild-type and synonymously barcoded viral strains are in \SUPPFILE{plasmids}.
These genes were cloned into the pHW2000~\citep{hoffmann2000dna} reverse-genetics plasmid.

Both viral strains were generated by reverse genetics using the pHW18* series of bi-directional plasmids~\citep{hoffmann2000dna}.
We controlled the durations and MOI during viral passaging since these factors can greatly affect the accumulation of defective viral particles~\citep{xue2016propagation}.
The viruses were generated by reverse genetics in co-cultures of 293T and MDCK-SIAT1 cells in influenza growth media (Opti-MEM supplemented with 0.01\% heat-inactivated FBS, 0.3\% BSA, 100 U of penicillin/ml, 100 $\mu$g of streptomycin/ml, and 100 $\mu$g of calcium chloride/ml) and then propagated in MDCK-SIAT1 cells in influenza growth media using the same basic procedures detailed in \citep{russell2018extreme}.
Specifically, after generation by reverse genetics, the wild-type variant was expanded at an MOI of 0.001 for 72 hours twice in MDCK-SIAT1 cells, and the synonymously barcoded variant was expanded once at an MOI of 0.01 for 60 hours.
The MOIs for this passaging are based on titers determined using TCID50 assays via the formula of Reed and Muench~\citep{reed1938simple} as implemented at \url{https://github.com/jbloomlab/reedmuenchcalculator}.

\subsection*{Flow cytometry analyses for HA expression}
For the single-cell experiments (which only examine the transcriptional results of a single cycle of infection), we were most interested in the titer of viral particles that are transcriptionally active for a single round of infection of A549 cells.
We estimated titers of transcriptionally active virions by staining for HA expression in virus-infected A549 cells.
Specifically, we infected A549 cells (or one of the A549 reporter cell line variants as indicated) in influenza growth medium, and at 13 to 14 hours post-infection, we trypsinized cells, re-suspended in phosphate-buffered saline (PBS) supplemented with 2\% heat-inactivated fetal bovine serum (FBS), and stained with 10 $\mu$g/ml of H17-L19, a mouse monoclonal antibody previously shown to bind to the HA from the A/WSN/1933 strain of virus~\citep{doud2017complete}.
After washing in PBS supplemented with 2\% FBS, the cells were stained with a goat anti-mouse IgG antibody conjugated to APC, washed, fixed in 1\% formaldehyde in PBS, washed again, and then analyzed by flow cytometry to determine the fraction expressing detectable HA protein.

\subsection*{Single-cell transcriptomics of IFN-enriched infected cells using 10X Chromium}
The single-cell transcriptomics and virus sequencing was performed using the A549 cells with the \textit{IFNB1} LNGFR$\Delta$C-P2A-mNeonGreen reporter.
A schematic of the experiment is shown in \FIG{workflow}.

The wild-type and synonymously barcoded viruses were mixed with the goal of adding equal numbers of transcriptionally active HA-expressing virions of each virus strain.
The cells were then infected with this mixture at a dose designed to infect about half the cells (\FIG{transcriptomics}C suggests that the actual rate of detectable infection was slightly lower).
Infections were allowed to proceed for 12 hours.
The cells were then trypsinized, the trypsin was quenched with D10 media, and cells were resuspended in de-gassed PBS supplemented with 0.5\% bovine serum albumin and 5 mM EDTA. 
To enrich IFN+ cells, the cells were then incubated with anti-LNGFR MACSelect Microbeads (Miltenyi Biotec) and twice passed over an MS magnetic column (Miltenyi Biotec), retaining the bound (and presumably IFN-enriched) population each time. 
This MACS sorting is expected to give approximately the enrichment for IFN+ cells shown in \SUPPFIG{MACS}.
The original, unsorted, population was then added back in to $\sim$10\% of the final cell fraction in order to ensure the presence of interferon negative cells. 
At this point, uninfected canine (MDCK-SIAT1) cells were also added to $\sim$5\% of the final cell fraction to enable quantification of the cell multiplet rate (\FIG{transcriptomics}A) and background viral mRNA in uninfected cells (\FIG{transcriptomics}C). 
We began this entire process of cell collection and enrichment at 12 hours post-infection, but the process (which was performed at room temperature) took about an hour, and thus we consider the cells to have been analyzed at 13 hours post-infection.
The final cell suspension was counted using a disposable hemocytometer and loaded on the 10x Genomics Chromium instrument~\citep{zheng2017massively}, targeting capture of $\sim$1,500 cells. 

This sample was then processed to create libraries for Illumina 3'-end sequencing according to the 10X Genomics protocol using the Chromium Single Cell 3' Library and Gel Bead Kit v2 with one important modification: rather than process all full-length cDNA through enzymatic fragmentation, several nanograms were retained for targeted full-length viral cDNA sequencing as described below.
The single-cell transcriptomics library was sequenced on an Illumina HiSeq 2500, and the data analyzed as described below.


\subsection*{Enrichment and preparation of viral cDNA for PacBio sequencing}
We amplified virus-derived molecules from the small amount of cDNA retained from the 10X Genomics protocol for PacBio sequencing of the full-length cDNA.
All molecules in this cDNA have at their 3' end the cell barcode and UMI plus the constant adaptor sequence that is added during the 10X protocol (see \FIG{workflow} for simple schematic, or the detailed analysis notebook at \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/blob/master/pacbio_analysis.ipynb} or \SUPPFILE{pacbio_analysis} for at detailed schematic).
However, we only wanted to PacBio sequence cDNA molecules derived from virus, since it would be prohibitively expensive to sequence all molecules.
We therefore needed to enrich for the viral molecules, a process made challenging by the need to also retain the 10X adaptor / UMI / cell barcode at the 3' end (the primer at the 5' end can be specific, but the one at the 3' end must bind the common 10X adaptor shared among all molecules).

We first performed a multiplex PCR reaction on 1 ng of the full-length 10X cDNA using a 3' primer complementary to the common 10X adaptor, and a multiplex mix of eight 5' primers, one specific for the mRNAs from each of the eight viral gene segments (although some viral segments produce multiple splice forms, all mRNAs from a given segment share the same 5' end).
The sequences of these primers are in \SUPPFILE{primer_sequences}.
A major concern during these PCRs is strand exchange (see \SUPPFIG{PacBio_supp}B) which would scramble the cell barcodes and mutations on viral cDNAs.
To reduce strand-exchange and hopefully obtain more even PCR amplification across segments, we performed emulsion PCRs using the Micellula DNA Emulsion Kit (Roboklon).
Emulsion PCR involves encapsidating DNA template molecules in a reverse-phase emulsion, with each template positive droplet serving as a microreactor.
This process physically separates disparate template molecules, preventing strand exchange and allowing each molecule to be amplified to exhaustion of its droplet's reagents without competing with the broader pool of PCR-amenable molecules~\citep{boers2015micelle}.
We performed the PCRs using Kapa HiFi Hotstart ReadyMix, supplementing the reactions with additional BSA to a final concentration of 0.1 mg/ml and using a volume 100$\mu$l .
Both the common 3' primer and the multiplex mix of eight 5' primers were added to a final concentration of 0.5$\mu$M.
We performed 30 cycles of PCR, using an extension time of 2 minutes 15 seconds at 67$^{\circ}$C, and a melting temperature of 95$^{\circ}$C.
This melting temperature is lower than the standard 98$^{\circ}$C melting step suggested by the manufacturer for Kapa HiFi because we wanted to avoid collapse of emulsion integrity at high temperature.
We performed 30 cycles in order to saturate the reagents in each emulsion---unlike for open PCR reactions, the physical occlusion of amplified material in emulsion PCR reduces artifacts at high cycle numbers.

Because were were concerned that the multiplex PCR would still result in highly uneven amplification of different influenza cDNAs due to differences in their expression levels and lengths, the product of this multiplex PCR was subjected to eight additional individual emulsion PCR reactions, each using only a single segment-specific 5' primer as well as the common 3' primer, using 1 ng of material in each reaction.
The material from these eight segment-specific PCRs was then pooled with the goal of obtaining a equimolar ratio of segments, and sequenced on one SMRT Cell in a PacBio RS II and one SMRT Cell of a PacBio Sequel. 
Detailed results from the analysis of these first two sequencing runs is shown in the PacBio analysis notebook at \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/blob/master/pacbio_analysis.ipynb} and \SUPPFILE{pacbio_analysis}.
These results showed that although the PCRs substantially enriched for influenza molecules, the relative coverage of the different viral genes was still very uneven, with the longer genes (especially the polymerase genes) severely under-sampled.

To try to improve coverage of the polymerase genes, we produced two new sequencing pools: one consisting of the five shortest viral segments (HA, NP, NA, M, and NS) from the aforementioned segment-specific emulsion PCRs, and the other consisting of the three longer polymerase segments (PB2, PB1, and PA).
The former was sequenced on one cell of a single SMRT Cell of a PacBio Sequel, and the latter on two additional SMRT Cells of a PacBio Sequel. 
As is shown in the PacBio analysis notebook at \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/blob/master/pacbio_analysis.ipynb} (see also \SUPPFILE{pacbio_analysis}), even with these new sequencing runs the coverage remained relatively low for the polymerase genes---and most of the reads we did obtain were dominated by shorter internally deleted variants of the polymerase genes, which arise commonly during influenza replication~\citep{xue2016propagation} and are presumably preferentially amplified during PCR.

To obtain more reads for longer full-length polymerase variants, we therefore subjected 10 ng of our amplified material for each polymerase segment to a bead selection using SPRIselect beads at a volume ratio of 0.4 to select for larger molecules. 
This selection removes most low-molecular weight DNA species including internally-deleted defective segments.
Material from this selection was then amplified using 16 (PB1) or 14 (PB2 and PA) cycles of a non-emulsion PCR using the standard conditions recommended by the Kapa HiFi Hotstart ReadyMix (extension at 67$^{\circ}$C for 2 minutes 15 seconds, and melting at 98$^{\circ}C$).
The use of relatively few PCR cycles was designed to prevent the occurrence of the artifacts (including strand exchange) that occur in non-emulsion PCRs as the reactions approach saturation.
We pooled the products of these reactions from this size-selection and sequenced on a SMRT Cell of a PacBio Sequel.
As is shown in the PacBio analysis notebook at \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/blob/master/pacbio_analysis.ipynb} (see also \SUPPFILE{pacbio_analysis}), this sequencing yielded modestly more full-length polymerase variants, but they were still heavily undersampled compared to other viral genes.

To further to improve recovery of full-length PB1, PB2, and PA, we therefore took an alternative approach that allowed us to perform a specific PCR for full-length polymerase variants.
Specifically, we circularized the template molecules (including the full length of genes plus the cell barcode, UMI, and 10X adaptor), and then used two segment-specific primers that annealed in apposition near the center of each polymerase gene to linearize these circular molecules.
Only molecules that contain the middle of the polymerase genes (which are typically full-length) are linearized by this process.
In the downstream computational analysis, we can then determine the full sequence of the gene as well as the cell barcode of the initial molecule from which the linearized molecule is derived.
Specifically, we first used 2.5 ng of our already-amplified segment-specific material in a 10-cycle PCR to append circularization adapters (see \SUPPFILE{primer_sequences} for sequences), and cleaned the resultant mixture using SPRIselect beads at a volume ratio of 0.4.
We then used 10 ng of this amplified material in a 20$\mu$l NEBuilder reaction using an extended reaction time of 50 minutes in order to circularize the molecules.
We next incubated these reactions for 1 hour at 37$^{\circ}$C with exonuclease V and additional ATP to a final increase in concentration of 1 mM to digest all non-circularized molecules.
The circularized and digested material was then cleaned using SPRIselect beads at a volume ratio of 0.4.
This material was then used as template for three non-emulsion PCRs specific to PB2, PB1, or PA, using two segment-specific primers that align to the central portion of each gene but in apposition to each another (see \SUPPFILE{primer_sequences} for sequences).
These linearization reactions used 20 (PB2) or 26 (PB1 and PA) PCR cycles, and the resulting products were cleaned using SPRIselect beads at a volume ratio of 1.0.
This material was pooled to produce an equimolar mixture of full-length PB1, PA, and PB2 and sequenced in an additional SMRT Cell of PacBio Sequel. 
As is shown in the PacBio analysis notebook at \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/blob/master/pacbio_analysis.ipynb} (see also \SUPPFILE{pacbio_analysis}), this process efficiently yielded many full-length polymerase variants.

The computational analyses of the full-length viral gene sequences described below used the combination of the data from all of these reactions.
The number of sequences obtained for each gene after pooling the data from all reactions is shown in \SUPPFIG{PacBio_supp}A, which also indicates that the net rate of strand exchange is very low (see \SUPPFIG{PacBio_supp}B for an illustration of how this is determined).
A more detailed breakdown of the coverage of each gene and data showing a consistently low rate of strand exchange for all PacBio runs is at \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/blob/master/pacbio_analysis.ipynb} (see also \SUPPFILE{pacbio_analysis}).
Importantly, the various PCR biases and enrichment schemes mean that the coverage of molecules by the PacBio sequencing is not proportional to their original abundance in the starting mRNA.
However, as described in the computational analysis section below, the final analyses use the cell barcodes and UMIs in conjunction with the standard 10X Illumina sequencing to ensure that none of the conclusions are affected by the disproportionate amplification of some molecules during the PacBio library preparation (for instance, duplicate UMIs are removed from the PacBio data, and all conclusions about gene abundance or absence are based on the Illumina data).

\subsection*{qPCR for viral genes and IFN}
We performed qPCR on reverse-transcribed mRNA for influenza HA (to quantify viral transcription), IFNB1 (to quantify IFN induction), and L32 (a cellular housekeeping gene for normalization).
For the qPCR, we used the SYBR Green PCR Master Mix (Thermo Fisher) according to the manufacturer's protocol using oligo-dT primers.
The qPCR primers were: HA primer 1, 5'-\texttt{GGCCCAACCACACATTCAAC}-3'; HA primer 2, 5'-\texttt{GCTCATCACTGCTAGACGGG}-3'; IFNB1 primer 1, 5'-\texttt{AAACTCATGAGCAGTCTGCA}-3'; IFNB1 primer 2, 5'-\texttt{AGGAGATCTTCAGTTTCGGAGG}-3'; L32 primer 1, 5'-\texttt{AGCTCCCAAAAATAGACGCAC}-3'; L32 primer 2, 5'-\texttt{TTCATAGCAGTAGGCACAAAGG}-3'. 

For the qPCR in \SUPPFIG{delqPCR}, A549 cells were seeded at a density of $10^4$ cells/well in a 96-well plate in D10 media 24 hours prior to infection, with four independent wells seeded per experimental treatment. 
Immediately prior to infection D10 media was removed and replaced with influenza growth media and infected with the indicated influenza strains at a MOI of 0.4 based on TCID50 in MDCK-SIAT1 cells.
For the cells with cycloheximide added to block protein expression (and hence secondary transcription), cycloheximide was added to a final concentration of 50 $\mu$g/ml (a concentration sufficient to block secondary transcription~\citep{killip2014activation}) at the time of infection.
After 8 hours, mRNA was harvested using the CellAmp Direct RNA Prep Kit for RT-PCR, reverse-transcribed using an oligo-dT primer, and qPCR was performed as described above.

\subsection*{Viruses and experiments for validation experiments}
In \FIG{validation}, we tested the IFN inducing capacity of a variety of viral mutants identified in the single-cell experiments.
For point-mutant viruses, we created variants for all amino-acid substitutions found in PB1 and NS among IFN+ cells that did not also lack NS.
One of these mutants (amino-acid substitution S704P in PB1) did not reach sufficient titers in a single attempt to generate it by reverse genetics, and so was dropped from the experiment (note that we did not attempt replicates of the reverse genetics for this mutant, and so are \emph{not} confident in drawing strong conclusions about its actual attenuation).
This left six point-mutant viruses: four with point mutations in PB1, and two with point mutations in NS.
We also created a mutant virus that contained the internal deletion in PB1 found in an IFN+ cell.
In addition, we created a virus with an inactivated NS1 to mimic the infections that failed to express NS (we were unable to use complementing cells to generate a viral stock that completely lacked the NS segment).
This NS1stop virus contained six nucleotide changes resulting in the addition of five in-frame stop codons in NS1 starting 10 nucleotides downstream of the 5' splice donor site, thereby disrupting NS1 while leaving NS2 (NEP) intact.
All of these mutants were cloned into the pHW2000 bi-directional reverse-genetics plasmid~\citep{hoffmann2000dna} in order to enable generation of viruses encoding the mutant genes.
 \SUPPFILE{plasmids} provides the full sequences for all of these plasmids.

We generated the wild-type and point-mutant viruses for the validation experiments in \FIG{validation}A by reverse genetics using the pHW18* series of WSN reverse genetics plasmids~\citep{hoffmann2000dna}, but substituting the appropriate mutant plasmid listed in \SUPPFILE{plasmids} for the wild-type plasmid for that gene.
To generate the viruses from these plasmids, we transfected an equimolar mix of all eight plasmids into co-cultures of 293T and MDCK-SIAT1 cells seeded at a ratio of 8:1.
At 24 hours post-transfection, we changed media from D10 to influenza growth media. 
At 50 hours post-transfection (for the replicate 1 viruses in \SUPPFIG{validationflow}A) or 72 hours (for the replicate 2 viruses in \SUPPFIG{validationflow}), we harvested the virus-containing supernatant, clarified this supernatant by centrifugation at 300$\times$g for 4 min, and stored aliquots of the clarified viral supernatant at -80$^{\circ}$C.
We then thawed aliquots and titered by TCID50 on MDCK-SIAT1 cells.
For the infections in \SUPPFIG{validationflow}A, we wanted to use equivalent particle counts, so we normalized all viruses to an equivalent hemagglutination titer on turkey red blood cells~\citep{hirst1942quantitative}.
 Briefly, a solution of 10\% v/v red blood cells (LAMPIRE Biological Laboratories, Fisher Scientific catalogue number 50412942)  was washed in PBS and diluted to a final concentration of 0.5\% v/v. 
Two-fold serial dilutions of virus were added to an equal volume of diluted red blood cells, and titer was measured as the highest dilution of viral stock at which complete hemagglutination of red blood cells was observed.
We then performed infections of the A549 reporter cell line at equivalent hemagglutination titer and analyzed the data as described in \SUPPFIG{validationflow}A. 

To generate the NS1stop mutant virus and the wild-type and PB1del385to2163 mutant viruses in \SUPPFIG{validationflow}B, we used slightly different procedures.
The wild-type virus was generated by reverse genetics as described for the point-mutant viruses above, harvested at 48 hours post-transfection, and then passaged on MDCK-SIAT1 cells for 36 hours at an MOI of 0.05---conditions that we previously validated to lead to relatively little accumulation of defective particles~\citep{russell2018extreme}.
The NS1stop virus was similarly generated, but was passaged for 48 rather than 36 hours, since it had slower growth kinetics and so needed a longer period of time to reach high titers.
The viruses with deletions in the PB1 segment could not be generated in normal 293T and MDCK-SIAT1 cells, since they required the exogenous expression of the PB1 protein.
Therefore, these viruses were generated in previously described 293T and MDCK-SIAT1 cells that had been engineered to constitutively express PB1~\citep{bloom2010permissive}.
These viruses were harvested from transfections at 72 hours, and passaged twice in the MDCK-SIAT1 cells constitutively expressing PB1 at a MOI of 0.001 for 72 hours and 0.01 for 48 hours. 
This passaging was necessary as viral titers from transfections were too low to generate sufficient virus from a single passage.
The wild-type and NS1stop viruses were titered by TCID50 on MDCK-SIAT1 cells, and the PB1 deletion viruses were titered on the MDCK-SIAT1 cells constitutively expressing PB1.
The infections in \SUPPFIG{validationflow}B were performed at equivalent TCID50s as described in the legend to that figure.
That these equivalent TCID50s were also roughly equivalent in terms of particles capable of undergoing primary transcription is shown in \SUPPFIG{delqPCR}.

\subsection*{Computational analysis of single-cell transcriptomic and viral sequence data}
A computational pipeline that performs all steps in the data analysis is available at \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell}. 
This pipeline is orchestrated by \texttt{Snakemake}~\citep{koster2012snakemake}, and begins with the raw sequencing data and ends by generating the figures shown in this paper.
The sequencing data and annotated cell-gene matrix are available on the GEO repository under accession GSE120839 (\url{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE120839}).

Briefly, the raw deep sequencing data from the Illumina 3'-end sequencing were processed using the 10X Genomics software package \texttt{cellranger} (version 2.2.0). 
We built a multi-species alignment reference consisting of a concatenation of the human and influenza virus transcriptomes (the first ``species'') and the canine transcriptome (the second ``species''). 
The human transcriptome was generated by filtering genome assembly GRCh38 for protein-coding genes defined in GTF file GRCh38.87.
The influenza virus transcriptome consisted of the mRNAs for the wild-type A/WSN/1933 virus strain in \SUPPFILE{plasmids} (the \texttt{cellranger} alignment is sufficiently permissive that it aligns sequences from both the wild-type and synonymously barcoded viral variants to this transcriptome).
The canine transcriptome was generated by filtering genome assembly CanFam3.1 for protein-coding genes defined in GTF file CanFam3.1.87.
The \texttt{cellranger} software was used to align the Illumina 3'-end sequencing reads to this multi-species transcriptome, call human+influenza and canine cells (\FIG{transcriptomics}A), and generate a matrix giving the expression of each gene in each single cell.
We used a custom Python script to determine the number of influenza virus reads that could be assigned to the wild-type or synonymously barcoded virus, and added this information to the annotated the cell-gene matrix.

The PacBio sequences of the full-length viral genes were analyzed as follows.
First, we used version 3.1.0 of PacBio's \texttt{ccs} program (\url{https://github.com/PacificBiosciences/unanimity}) to build circular consensus sequences (CCSs) from the subreads files, requiring at least 3 passes and a minimum accuracy of 0.999.
We further processed these CCSs using custom Python code and the \texttt{minimap2}~\citep{li2018minimap2} long-read aligner (version 2.11-r797).
The Python code has been implemented in the API of \texttt{dms\_tools2} (\url{https://jbloomlab.github.io/dms_tools2/}~\citep{bloom2015software}) package (version 2.3.0).
A Jupyter notebook that performs these analyses is at \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/blob/master/pacbio_analysis.ipynb}, and is also provided in HTML form as \SUPPFILE{pacbio_analysis}.
We refer the reader to this notebook for a detailed description and extensive plots showing the results at each step.
Here is a brief summary: we filtered for CCSs that had the expected 5' termini (from the influenza-specific primers) and 3' termini (corresponding to the 10X adaptor), and for which we could identify the cell barcode, UMI, and polyA tail.
We aligned the cDNAs flanked by these termini to the influenza transcriptome, and performed a variety of quality control steps.
At this point, we examined whether cDNAs had the synonymous viral barcodes at both ends or neither end as expected in the absence of strand exchange, and reassuringly found that strand exchange was rare (\SUPPFIG{PacBio_supp}).
The small number of CCSs with identifiable strand exchange were filtered from further analysis.
We then further filtered for CCSs that contained valid cell barcodes as identified by the \texttt{cellranger} pipeline, and kept just one CCS per UMI (preferentially retaining high-quality CCSs that aligned to full-length cDNAs).
We then removed from the CCSs the barcoding synonymous mutations that we had engineered into one of the two viral variants.
Finally, we used the CCSs to call the sequence of the viral gene in each cell, calling mutations separately for each viral barcode variant.
We called mutations (insertions, deletions, and substitutions) in the viral gene sequences as follows:
\begin{enumerate}
\item Mutations with accuracies less than 0.999 (which constitute $<$0.5\% of all mutations) were ignored.
\item If all CCSs for a particular viral-barcode variant of a gene in a cell were wild-type, it was called as wild type.
\item If any CCSs for a particular viral-barcode variant of gene in a cell had a mutation, then require at least two CCSs to call the sequence.
\item If at least two and $>$30\% of the CCSs had a specific mutation, then call that mutation as present and note its frequency among the CCSs. The exception was single-nucleotide indels in homopolymers, for which we required three CCSs to call a mutation (the reason is that the main mode of PacBio sequencing errors is short indels in homopolymers).
\end{enumerate}
The plots in \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/blob/master/pacbio_analysis.ipynb} or \SUPPFILE{pacbio_analysis} indicate that these are reasonable mutation-calling criteria.
We could call the sequences of all expressed viral genes in about half of the infected cells (\SUPPFIG{ncells}).
The mutations called using this pipeline are shown in \FIG{genotypes}, and \SUPPFILE{genotypes} gives the number of CCSs supporting each mutation call.
The called sequences of the viral genes were added to the annotated cell-gene matrix.

Finally, we process the annotated cell-gene matrix in R to generate the plots shown in this paper.
This analysis utilized a variety of R and Bioconductor~\citep{huber2015orchestrating} packages, including \texttt{Monocle}~\citep{qiu2017reversed, trapnell2014dynamics} and \texttt{ggplot2}.
A Jupyter notebook that performs these analyses is at \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/blob/master/monocle_analysis.ipynb}, and is also provided in HTML form as \SUPPFILE{monocle_analysis}.
We refer the reader to this notebook for a detailed description and a variety of additional plots not included in the paper.
Briefly, we first filtered cells that were extreme outliers in the amount of mRNA as shown in \FIG{transcriptomics}B.
We used the uninfected canine cells to estimate the percentage of total mRNA in a cell that would come from influenza purely due to background (e.g., from cell lysis) in the absence of infection, and called as infected the human cells for which significantly more than this amount of mRNA was derived from influenza under a Poisson model (\FIG{transcriptomics}C).
We next used a Poisson model parameterized by the amount of expected background mRNA for each influenza gene to call the presence or absence of each influenza gene in each infected cell (\FIG{transcriptomics}D and \SUPPFIG{transcriptomics_supp}A). 
To identify cells that were co-infected with both viral barcodes (\FIG{transcriptomics}F), we used a binomial test to identify cells for which we could reject the null hypothesis that at least 95\% of viral mRNA was derived from the more common viral barcode.
We called IFN+ and ISG+ cells using the heuristic thresholds shown in \FIG{transcriptomics}G and \SUPPFIG{transcriptomics_supp}C,D, respectively.
We counted IFN mRNAs as any IFN-$\alpha$, IFN-$\beta$, or IFN-$\lambda$ transcripts, since type I and type III IFN were highly correlated (\SUPPFIG{transcriptomics_supp}B).
We counted ISG mRNAs as any of CCL5, IFIT1, ISG15, or Mx1.
The plot in \FIG{genotypes} summarizes all of the genotypic information, and was created in substantial part using \texttt{gggenes} (\url{https://github.com/wilkox/gggenes}).

\clearpage
\nolinenumbers
\normalsize

\section*{SUPPORTING INFORMATION}

\begin{suppfig}
\begin{minipage}[t]{0.46\textwidth}
{\bf \Large A}
\includegraphics[width=0.9\textwidth, valign=t]{figures/IFN_stochastic/IFN_reporter/Sendai_validation.pdf}
\end{minipage}
\begin{minipage}[t]{0.55\textwidth}
{\bf \Large B}
\includegraphics[width=0.57\textwidth, valign=t]{figures/IFN_stochastic/IFN_reporter/IFNbeta_IFN_lambda_correlated.pdf}
\vspace{0.1in}

{\bf \Large C}
\includegraphics[width=0.9\textwidth, valign=t]{figures/IFN_stochastic/Flow/flow_plot.pdf}

\end{minipage}

\caption{
{\bf Validation of IFN reporter, related to \FIG{IFNrare}.}
{\bf (A)}
IFN reporter cell lines were infected at high MOI with the Cantell strain of Sendai virus~\citep{strahle2006sendai}.
The name of the reporter is indicated above the plots.
At 13 hours post-infection, activation of the reporter was monitored by flow cytometry using the marker indicated below each plot (a fluorescent protein or antibody staining for the LNGFR$\Delta$C using PE-conjugated anti-LNGFR antibody from Miltenyi Biotec).
Sendai infection efficiently activated the IFN reporter in all cases.
{\bf (B)}
The type I and type III IFN reporters are highly correlated in their activation.
An A549 cell line was generated by transduction with both IFN-$\beta$ and IFN-$\lambda$ reporters driving mCherry and ZsGreen, respectively.
The cells were infected with two different stocks of WSN influenza, or stocks with a deletion in PB1 or stop codons in NS1.
After 13 hours, cells were analyzed by flow cytometry.
Cells positive for either reporter were further analyzed.
As shown in the FACS plots, expression of the \textit{IFNB1} and \textit{IFNL1} reporters is highly correlated.
{\bf (C)}
Flow cytometry data for \FIG{IFNrare}B.
A549 cells with the \textit{IFNL1} reporter driving LNGFR$\Delta$C-ZsGreen were not infected, infected with saturating amounts of the Cantell strain of Sendai virus, or infected at a target MOI of 0.3 with the stock of influenza virus used in the single-cell experiment.
At 13 hours, cells were analyzed by FACS for HA and ZsGreen expression.
Each condition was done in triplicate.
Contour plots show density of all cells, and IFN+ cells are also indicated by orange dots.
Cells were classified as HA+ or IFN+ based on gates set to put 0.05\% of the uninfected cells in these populations.
For influenza-infected cells, the percentage IFN+ was calculated only among HA+ cells (since these are the ones that are infected).
For uninfected and Sendai-infected cells, the percentage IFN+ was calculated among all cells, since the cells do not express HA.
}
\label{suppfig:reportervalidation}
\end{suppfig}

\begin{suppfig}
\includegraphics[width=\textwidth]{figures/MACS/MACS.pdf}
\caption{
{\bf Example MACS enrichments of IFN+ influenza-infected cells, related to \FIG{workflow}.}
A549 cells with the \textit{IFNB1} LNGFR$\Delta$C-mNeonGreen reporter were infected with wild-type WSN influenza (two different viral stocks) at a target MOI of 0.1 TCID50 per cell.
After infection had proceeded for 12 hours, the cells were twice magnetically sorted for LNGFR$\Delta$C expression over magnetic columns as detailed in the methods for the single-cell sequencing experiment.
{\bf (A)}
After sorting, the populations were analyzed by flow cytometry for IFN expression using the mNeonGreen fluorescent protein.
The plots show the distribution of fluorescence in the original population, the flow-through from the first column, and the MACS-sorted positive population after two columns.
As indicated by the percentages shown for the original and MACS-sorted population, this process led to substantial enrichment in IFN+ cells.
We expect that the IFN sorting for the actual single-cell sequencing led to similar enrichment, although we could not directly quantify this as the sorted cells in that case were immediately used for the sequencing and so could not be analyzed by flow cytometry.
{\bf (B)}
Analysis of expression of IFNB1 (relative to the housekeeping gene L32) by qPCR in the positive (IFN enriched) and negative (IFN depleted) populations from panel (A).
The qPCR validates a roughly 50- to 100-fold enrichment in total IFNB1 expression.
The qPCR was performed in quadruplicate (hence the four points for each sample).
}
\label{suppfig:MACS}
\end{suppfig}

\begin{suppfig}
\includegraphics[width=0.9\textwidth]{figures/single_cell_figures/p_cell_summary_supp.pdf}
\caption{
{\bf Additional plots from analysis of transcriptomic data, related to \FIG{transcriptomics}.}
{\bf (A)}
Fraction of infected cells that are called as expressing each viral gene.
The gray dashed line is at one.
Each gene is detected in $\sim$80-90\% of cells, roughly in line with prior estimates~\citep{brooke2013most, heldt2015single, dou2017analysis, russell2018extreme}.
The exception is NP, which is detected in virtually all infected cells.
The higher frequency of detecting NP could reflect a biological phenomenon, but we suspect it is more likely that cells lacking NP tend to have lower viral gene expression and so are not reliably called as infected in our experiments because the number of viral mRNAs is below the detection limit.
{\bf (B)}
Correlation between the fraction of cellular mRNA derived from type I and type III IFN in the A549 cells in our single-cell transcriptomics.
Each point represents one cell.
The plots are faceted by whether the cells are called as infected, and the Pearson correlation coefficient is shown.
{\bf (C)}
Distribution of ISGs expression across single infected and uninfected cells. 
For each cell, we quantified ISG expression as the total fraction of cellular mRNAs derived from four prototypical ISGs (IFIT1, ISG15, CCL5, and Mx1). 
We heuristically classify as ISG+ cells with $>10^{-3}$ of their cellular mRNA from these ISGs, and color these cells red.
Comparison to \FIG{transcriptomics}G shows that more cells are ISG+ than IFN+, probably because paracrine signaling can induce ISG expression in cells that do not themselves express IFN~\citep{stetson2006type,honda2006type}.
{\bf (D)} 
Correlation between the fraction of cellular mRNA derived from IFN and ISGs.
Each point is one cell, and the Pearson correlation coefficient is shown.
IFN and ISG expression are more correlated for infected than uninfected cells, probably because in the latter the ISG expression is more often due to paracrine signaling that does not induce expression of IFN itself.
Among both the infected and uninfected populations, there are many cells that express ISGs but not IFN, but no cells that express IFN but not ISGs.
}
\label{suppfig:transcriptomics_supp}
\end{suppfig}

\begin{suppfig}
{\bf \Large A}
\includegraphics[width=0.9\textwidth, valign=t]{figures/pacbio_single_cell_figures/ccs_per_viralbarcode.pdf}
\vspace{0.05in}

{\bf \Large B}
\includegraphics[width=0.52\textwidth, valign=t]{figures/StrandExchangeSchematic/StrandExchangeSchematic.pdf}
\caption{
{\bf PacBio CCS quality control, related to \FIG{genotypes}.}
{\bf (A)}
Number of PacBio CCSs that passed quality-control and aligned to each viral gene.
CCSs were obtained using several PacBio runs which were loaded with different amounts of each gene to balance coverage (see \SUPPFILE{pacbio_analysis}).
Therefore, unlike the transcriptomic data in \FIG{transcriptomics}, the numbers of CCSs is \emph{not} an indicator of a transcript's abundance.
Especially for the polymerase genes, many CCSs corresponded to genes with internal deletions, since these shorter forms are preferentially amplified by PCR.
Therefore, the plot is faceted by CCSs for any length of the gene, and for full-length genes.
The disproportionate sequencing of the shorter internally deleted genes does not affect the genotype calling since UMIs were used to collapse sequences from the same cDNA, and cell barcodes were used to collapse sequences from the same cell.
The bars in the plot are colored by whether the sequence is derived from the wild-type viral variant, the synonymously barcoded viral variant, or is a mixed-barcode molecule (see panel B).
We estimate~\citep{bloom2018estimating} that 5.7\% of molecules are chimeric due to PCR strand exchange.
About half of these chimeras can identified by the mixed viral barcodes and removed from subsequent analyses, leaving $\sim$3\% un-identified chimeras.
For some molecules (mostly polymerase genes with internal deletions) one of the barcode sites was deleted from the molecule and so the viral barcode could only be partially called.
{\bf (B)}
Strategy for detecting strand exchange in PacBio sequencing.
Preparation for PacBio sequencing required many PCR cycles, which can cause strand exchange that scrambles mutations and 10X cell barcodes / UMIs from different molecules.
If there is no strand exchange, all molecules are either wild-type or have the synonymous barcoding mutations at \emph{both} termini.
Strand exchange creates molecules with wild-type nucleotides at one termini and synonymous barcoding mutations at the other.
Panel A shows the frequencies of these molecules.
Since homologous recombination in influenza virus in negligible~\citep{boni2008homologous}, mixed-barcode molecules are \emph{not} generated naturally during co-infection.
}
\label{suppfig:PacBio_supp}
\end{suppfig}

\begin{suppfig}
\includegraphics[width=0.8\textwidth]{figures/single_cell_figures/p_cells_complete.pdf}
\caption{
{\bf Number of cells for which we could determine the full sequences of all genes expressed by the infecting virion(s).}
{\bf (A)} We could call the complete genotypes of the infecting virion(s) for the majority of cells infected with just a single viral barcode variant, but only a minority of cells co-infected with both viral barcodes.
{\bf (B)} The cells for which we could call complete viral genotypes tended to have higher expression of viral mRNA than cells for which we could not call complete genotypes.
Both facts makes sense.
Cells with more viral mRNA are more likely to have their viral cDNA captured in the PacBio sequencing, which is only captures a small fraction of the total transcripts identified by the 3'-end sequencing transcriptomic sequencing.
The lower calling rate for dual-barcode co-infections is probably because these co-infections have more viral genes that must be sequenced (potentially a copy of each viral gene from each viral variant), increasing the chances that one of these genes is missed by the PacBio sequencing. 
An important implication of this plot is that the cells for which we call complete viral genotypes are \emph{not} a random subsampling of all infected cells in the experiment, but are rather enriched for cells that have high levels of viral mRNA and do not have dual-barcode viral infections.
Note also that this plot is limited to the cells that were called as infected (\FIG{transcriptomics}C) and could clearly be classified as IFN- or IFN+ (\FIG{transcriptomics}G).
}
\label{suppfig:ncells}
\end{suppfig}

\begin{suppfig}
{\bf \Large A}\includegraphics[width=0.48\textwidth, valign=t]{figures/Validation_Figure/SNP_flow_plot.pdf}
\hspace{0.02\textwidth}
{\bf \Large B} \includegraphics[width=0.39\textwidth, valign=t]{figures/Validation_Figure/del_flow_plot.pdf}
\caption{
{\bf Flow cytometry data for \FIG{validation}.}
{\bf (A)}
Data for \FIG{validation}A.
A549 cells with the \textit{IFNL1} reporter driving LNGFR$\Delta$C-ZsGreen were infected with stocks of the indicated mutant.
After 13 hours, cells were stained for HA protein and analyzed by FACS.
Contour plots show density of all cells, and IFN+ cells are also indicated by orange dots.
Cells were classified as HA+ or IFN+ based on gates set to put 0.05\% of uninfected cells in these populations.
For infected cells, the percentage IFN+ was calculated among the HA+ cells (since these are the ones that are infected).
For uninfected cells, the percentage IFN+ was calculated among all cells, since uninfected cells do not express HA.
For each viral mutant, two independent stocks were assayed in duplicate (i.e., \#1a and \#1b are one viral stock, and \#2a and \#2b are the other).
The infections with replicate \#1 of the wild-type virus were performed at an MOI of 0.1 as determined by TCID50, and all other viruses were infected at an equivalent particle number as determined by HI assay. 
{\bf (B)}
Data for \FIG{validation}B.
The virus with the deletion in PB1 cannot be normalized by HA expression since it expresses less HA due to the lack of secondary transcription.
Therefore, all cells were infected at an equivalent MOI of 0.3 as determined by TCID50 on MDCK-SIAT1 cells for wild type and NS1stop, and on MDCK-SIAT1 cells expressing PB1~\citep{bloom2010permissive} for PB1del385to2163.
\SUPPFIG{delqPCR} shows that at these equivalent TCID50s, all variants had similar amounts of transcriptionally active virus in the absence of secondary transcription.
The percent IFN+ was calculated for \emph{all} cells (HA+ and HA-) since that is a more fair comparison for PB1del385to2163.
}
\label{suppfig:validationflow}
\end{suppfig}

\begin{suppfig}
\includegraphics[width=0.6\textwidth]{figures/Validation_Figure/qpcr_plot.pdf}
\caption{
{\bf Validation that the infections in \FIG{validation}B and \SUPPFIG{validationflow}B were performed at similar doses of virions capable of initiating primary transcription.}
In this experiment, A549 cells were infected at MOI of 0.4 (based on TCID50 as described in \SUPPFIG{validationflow}B), and then after 8 hours mRNA was harvested for qPCR on oligo-dT primed reverse transcription products.
The y-axis shows the ratio of viral HA mRNA to the housekeeping gene L32.
These infections were performed in the presence of absence of 50 $\mu$g/ml cycloheximide, which blocks protein synthesis and hence secondary transcription by newly synthesized viral proteins~\citep{killip2014activation}.
In the absence of cycloheximide, the viruses with deletions in PB1 produced less viral mRNA presumably because they could not produce PB1 protein for secondary transcription.
But in the presence of cycloheximide, all viruses produced similar amounts of viral mRNA, indicating that the dose of particles active for primary transcription is roughly equivalent across variants.
Each measurement was performed in quadruplicate.
}
\label{suppfig:delqPCR}
\end{suppfig}

\clearpage

\begin{suppfile}
\caption{{\bf Sequences of plasmids and viruses used in this study.}
The plasmid maps are for the IFN reporters in \FIG{IFNrare}A and the mutant viral genes cloned into the pHW* bi-directional reverse genetics plasmid~\citep{hoffmann2000dna}.
In addition, there are Genbank files giving the sequences of the wild-type and synonymously barcoded WSN viruses.
A README file explains the contents of the ZIP.
%See \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/tree/master/paper/figures/IFN_stochastic/IFN_reporter/plasmids}, \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/tree/master/paper/figures/FluVariantPlasmidMaps}, \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/blob/master/data/flu_sequences/flu-wsn.gb}, and \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/blob/master/data/flu_sequences/flu-wsn-double-syn.gb}
}
\label{suppfile:plasmids}
\end{suppfile}

\begin{suppfile}
\caption{
{\bf A text file giving the primers used to amplify the influenza cDNAs for PacBio sequencing.}
%See \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/tree/master/paper/figures/WorkflowSchematic/PacBio_primer_list.txt} 
}
\label{suppfile:primer_sequences}
\end{suppfile}

\begin{suppfile}
\caption{
{\bf A CSV file giving the genotypes in \FIG{genotypes}.}
%See \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/blob/master/paper/figures/single_cell_figures/genotypes.csv}.
}
\label{suppfile:genotypes}
\end{suppfile}

\begin{suppfile}
\caption{
{\bf A CSV file giving the viral mutations and related information in \FIG{mutations}.}
%See \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/blob/master/paper/figures/single_cell_figures/mutations.csv}.
}
\label{suppfile:mutations}
\end{suppfile}

\begin{suppfile}
\caption{\label{suppfile:pacbio_analysis}
{\bf HTML rendering of Jupyter notebook that analyzes the PacBio data.}
%See \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/raw/master/paper/figures/pacbio_single_cell_figures/pacbio_analysis.html}.
The actual Jupyter notebook is at \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/blob/master/pacbio_analysis.ipynb}.
}
\end{suppfile}

\begin{suppfile}
\caption{\label{suppfile:monocle_analysis}
{\bf HTML rendering of Jupyter notebook that analyzes the annotated cell-gene matrix.}
%See \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/raw/master/paper/figures/single_cell_figures/monocle_analysis.html}.
The actual Jupyter notebook is at \url{https://github.com/jbloomlab/IFNsorted_flu_single_cell/blob/master/monocle_analysis.ipynb}.
}
\end{suppfile}

\end{document}

